File: apps/frontend/src/lib/api/ai.ts
import apiClient from './client';

export interface AIInteractionCreate {
  question: string;
  context_type: string; // 'course', 'material', 'quiz'
  context_id: string;
}

export interface AIInteraction {
  id: string;
  user_id: string;
  question: string;
  answer: string;
  context_type: string;
  context_id: string;
  created_at: string;
}

export interface AISearchQuery {
  query: string;
  context_type: string; // 'course', 'material'
  context_id: string;
  limit?: number;
}

export interface AISearchResult {
  results: Array<{
    id: string;
    content: string;
    similarity: number;
    source_id: string;
    source_type: string;
    metadata: Record<string, any>;
  }>;
}

export interface AIQuizGenerationRequest {
  material_id: string;
  num_questions: number;
  difficulty: string; // 'easy', 'medium', 'hard'
}

export interface AIExplanationRequest {
  content_id: string;
  learning_style?: string; // 'visual', 'auditory', 'reading', 'kinesthetic'
  difficulty_level?: string; // 'beginner', 'intermediate', 'advanced'
}

export interface LearningStyleUpdate {
  visual_score: number; // 1-10
  auditory_score: number; // 1-10
  reading_score: number; // 1-10
  kinesthetic_score: number; // 1-10
}

const aiService = {
  async getLearningStyle() {
    const response = await apiClient.get('/ai/learning-style');
    return response.data;
  },

  async updateLearningStyle(data: LearningStyleUpdate) {
    const response = await apiClient.put('/ai/learning-style', data);
    return response.data;
  },

  async getLearningRecommendations() {
    const response = await apiClient.get('/ai/learning-recommendations');
    return response.data;
  },

  async getAIUsageMetrics(timePeriod = 'week') {
    const params = { time_period: timePeriod };
    const response = await apiClient.get('/ai/usage-metrics', { params });
    return response.data;
  },

  async getAIPerformanceMetrics(timePeriod = 'week') {
    const params = { time_period: timePeriod };
    const response = await apiClient.get('/ai/performance-metrics', { params });
    return response.data;
  },

  async processMaterialForAI(materialId: string) {
    const response = await apiClient.post(`/ai/process-material/${materialId}`);
    return response.data;
  },

  async generateCourseQuiz(courseId: string, numQuestions = 10, difficulty = 'medium') {
    const params = { num_questions: numQuestions, difficulty };
    const response = await apiClient.post(`/ai/generate-course-quiz/${courseId}`, null, { params });
    return response.data;
  },

  async askAI(data: AIInteractionCreate) {
    const response = await apiClient.post('/ai/ask', data);
    return response.data;
  },

  async semanticSearch(data: AISearchQuery) {
    const response = await apiClient.post('/ai/search', data);
    return response.data;
  },

  async generateQuiz(data: AIQuizGenerationRequest) {
    const response = await apiClient.post('/ai/generate-quiz', data);
    return response.data;
  },

  async explainContent(data: AIExplanationRequest) {
    const response = await apiClient.post('/ai/explain', data);
    return response.data;
  },
};

export default aiService;

File: apps/api/app/schemas/ai.py
from typing import List, Optional, Dict, Any
from datetime import datetime
from pydantic import BaseModel, Field, UUID4
from enum import Enum

class AIModelType(str, Enum):
    """AI model type enumeration."""
    GPT_3_5 = "gpt-3.5-turbo"
    GPT_4 = "gpt-4"
    CUSTOM = "custom"

class AIInteractionType(str, Enum):
    """AI interaction type enumeration."""
    QUESTION = "question"
    EXPLANATION = "explanation"
    QUIZ_GENERATION = "quiz_generation"
    SUMMARY = "summary"

class AIInteractionBase(BaseModel):
    """Base AI interaction schema."""
    query: str
    interaction_type: AIInteractionType = AIInteractionType.QUESTION
    context_ids: Optional[List[UUID4]] = None  # IDs of related content chunks
    model: AIModelType = AIModelType.GPT_3_5

class AIInteractionCreate(AIInteractionBase):
    """AI interaction creation schema."""
    user_id: UUID4
    course_id: Optional[UUID4] = None
    material_id: Optional[UUID4] = None

class AIInteraction(AIInteractionBase):
    """AI interaction response schema."""
    id: UUID4
    user_id: UUID4
    course_id: Optional[UUID4] = None
    material_id: Optional[UUID4] = None
    response: str
    tokens_used: int
    created_at: datetime
    
    class Config:
        orm_mode = True

class AISearchQuery(BaseModel):
    """AI semantic search query schema."""
    query: str
    course_id: Optional[UUID4] = None
    limit: int = 5

class AISearchResult(BaseModel):
    """AI semantic search result schema."""
    content_chunk_id: UUID4
    content: str
    material_id: UUID4
    material_title: str
    similarity: float
    metadata: Dict[str, Any] = Field(default_factory=dict)
    
    class Config:
        orm_mode = True

class AIQuizGenerationRequest(BaseModel):
    """AI quiz generation request schema."""
    material_id: UUID4
    num_questions: int = 5
    question_types: List[str] = ["multiple_choice", "true_false"]
    difficulty: str = "medium"  # easy, medium, hard

class AIExplanationRequest(BaseModel):
    """AI explanation request schema."""
    content: str
    material_id: Optional[UUID4] = None
    detail_level: str = "medium"  # basic, medium, advanced

class LearningStyleUpdate(BaseModel):
    """Learning style update schema."""
    visual_score: Optional[int] = Field(None, ge=1, le=10)
    auditory_score: Optional[int] = Field(None, ge=1, le=10)
    reading_score: Optional[int] = Field(None, ge=1, le=10)
    kinesthetic_score: Optional[int] = Field(None, ge=1, le=10)
